### ISSUE TITLE
Need more `SighashCache` doc/API improvements

### ISSUE THREAD

**Issue by apoelstra** [rust-bitcoin/rust-bitcoin#1698]
We've been asked repeatedly (on IRC and recently [on StackExchange](https://bitcoin.stackexchange.com/questions/117465/in-a-spent-m-of-n-multisig-i-want-to-know-which-m-pubkeys-in-the-redeem-script)) how to compute transaction sighashes. Our current `SighashCache` is not very accessible.

There are two specific improvements we should make:

* Provide functionality to compute the `script_code` from input scriptpubkeys or addresses. In particular for segwit p2pkwh the scriptcode is surprising and weird.
* Provide a complete example of a sighash computation for a legacy, segwit and taproot pubkey transaction. Maybe for p2sh, p2wsh, and tapscript as well. (We may want to just write an examples.rs and link to it in the docs.)

**Comment by panicfarm** [comment]
Looking forward to it! Also would be helpful to see the examples for different sighash input flags: `SIGHASH_ALL`, `SIGHASH_SINGLE`, etc.

**Comment by Kixunil** [comment]
Heh, I think this is similar to what I complained about - stuff for cookbook.

**Comment by panicfarm** [comment]
@apoelstra I created a GitHub repo https://github.com/panicfarm/multisig-test/ with the examples. It's currently lacking P2TR examples. If needed, I can create a PR.

**Comment by apoelstra** [comment]
@panicfarm at first glance those look great! If you are offering to PR them to this repo, that would be very welcome.

**Comment by panicfarm** [comment]
> @panicfarm at first glance those look great! If you are offering to PR them to this repo, that would be very welcome.

@apoelstra 
Yes, I'd be happy to!  Is this the most appropriate dir https://github.com/rust-bitcoin/rust-secp256k1/tree/master/examples ?

Or maybe since these examples verify pubkeys/sigs, not just show how to compute sighashes , maybe it's better for https://github.com/rust-bitcoin/rust-bitcoin/tree/master/bitcoin/examples ?

**Comment by tcharding** [comment]
The examples depend on `rust-bitcoin` so they should be in https://github.com/rust-bitcoin/rust-bitcoin`tree/master/examples`.

**Comment by tcharding** [comment]
Or if you are feeling super game, you could patch the new cookbook

- https://rust-bitcoin.org/book/
- https://github.com/rust-bitcoin/www.rust-bitcoin.org/tree/master/cookbook

**Comment by panicfarm** [comment]
> @panicfarm at first glance those look great! If you are offering to PR them to this repo, that would be very welcome.

@apoelstra Submitted a PR https://github.com/rust-bitcoin/rust-bitcoin/pull/1776 . I am planning to add the P2TR examples soon. 

@tcharding I would be happy to contribute them to the book along with explanations, particularly when the P2TR examples are added. At present, the book appears to be empty.  Once you have written a chapter or two, I could contribute and match your desired writing style.

**Comment by tcharding** [comment]
No worries, thanks man.
